kevm-pyk/src/kevm_pyk/__main__.py

```diff
@@ -432,23 +458,95 @@ def _init_and_run_proof(claim_job: KClaimJob) -> tuple[bool, list[str] | None]:
 
             return passed, failure_log
 
-    topological_sorter = graphlib.TopologicalSorter(claims_graph)
-    topological_sorter.prepare()
-    with wrap_process_pool(workers=workers) as process_pool:
-        selected_results: list[tuple[bool, list[str] | None]] = []
-        selected_claims = []
-        while topological_sorter.is_active():
-            ready = topological_sorter.get_ready()
-            _LOGGER.info(f'Discharging proof obligations: {ready}')
-            curr_claim_list = [all_claim_jobs_by_label[label] for label in ready]
-            results: list[tuple[bool, list[str] | None]] = process_pool.map(_init_and_run_proof, curr_claim_list)
-            for label in ready:
-                topological_sorter.done(label)
-            selected_results.extend(results)
-            selected_claims.extend(curr_claim_list)
+    def worker(env: MyEnvironment) -> None:
+        while True:
+            msg = env.to_do_queue.get()
+            match msg:
+                case TodoQueueStop():
+                    return
+                case TodoQueueProofTask(claim_label):
+                    try:
+                        passed, failure_log = _init_and_run_proof(all_claim_jobs_by_label[claim_label])
+                        env.done_queue.put(
+                            DoneQueueTaskFinished(claim_label=claim_label, passed=passed, failure_log=failure_log)
+                        )
+                    except Exception as e:
+                        env.done_queue.put(
+                            DoneQueueTaskFinished(claim_label=claim_label, passed=False, failure_log=[str(e)])
+                        )
+
+    def coordinator(env: MyEnvironment, jobs_to_do: list[KClaimJob]) -> list[tuple[bool, list[str] | None]]:
+        id_to_job: dict[str, KClaimJob] = {job.claim.label: job for job in jobs_to_do}
+        remaining_job_labels: set[str] = {job.claim.label for job in jobs_to_do}
+        finished_job_labels: dict[str, tuple[bool, list[str] | None]] = {}
+
+        def get_a_ready_id() -> str | None:
+            for job_label in remaining_job_labels:
+                if set({d.claim.label for d in id_to_job[job_label].dependencies}).issubset(finished_job_labels.keys()):
```

Comment:
What if the dependency was finished in a prior job, and we just need to read from disk that it's the case? Will this catch that? It seems here that it will only work if it was proved _in this session_, not if it was proved in a prior session. Is that the case?

Reply:
I didn't think about that; you are probably right.